File: screenShotServer/sssEclipseProject/src/adshotrunner/powerpoint/PowerPointBackground.java
package adshotrunner.powerpoint;

import java.io.File;
import java.net.URL;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Timestamp;

import org.apache.commons.io.FileUtils;

import adshotrunner.errors.AdShotRunnerException;
import adshotrunner.system.ASRProperties;
import adshotrunner.utilities.ASRDatabase;

/**
 * Class to retrieve PowerPoint background information
 */
public class PowerPointBackground {

	//URL path for background and thumbnail images
	final public static String BACKGROUNDURLPATH = "http://s3.amazonaws.com/" + 
												    ASRProperties.containerForPowerPointBackgrounds() + "/";
	final public static String THUMBNAILURLPATH = "http://s3.amazonaws.com/" + 
												    ASRProperties.containerForPowerPointBackgrounds() + "/thumbnails/";
	
	//---------------------------------------------------------------------------------------
	//--------------------------------- Static Methods --------------------------------------
	//---------------------------------------------------------------------------------------	
	//***************************** Public Static Methods ***********************************
	/**
	 * Retrieves and returns a PowerPointBackground from the database with the provided ID.
	 * 
	 * @param backgroundID		ID of the PowerPointBackground
	 * @return					PowerPointBackground on success and NULL on failure
	 */	
	public static PowerPointBackground getPowerPointBackground(int backgroundID) {
		
		//Check to see if a PowerPointBackground with that ID exists in the database
		//Redundant check along with constructor
		try {
			ResultSet backgroundResults = ASRDatabase.executeQuery("SELECT * " + 
																   "FROM powerPointBackgrounds " + 
																   "WHERE PPB_id = " + backgroundID);
	
			//If a match was found, return the background 
			if (backgroundResults.next()) {
				return new PowerPointBackground(backgroundID);
			}	
			
			//Otherwise simply return null
			else {
				System.out.println("Unable to get PowerPointBackground with ID: " + backgroundID);
				return null;
			}
		}
		
		//If we couldn't query the database correctly or another error occurred, print it out and return null
		catch (Exception e) {
			e.printStackTrace();;
			return null;
		}		
	}

	//---------------------------------------------------------------------------------------
	//------------------------------------ Variables ----------------------------------------
	//---------------------------------------------------------------------------------------	
	//******************************** Private Variables ************************************
	/**
	 * ID of the PowerPoint background
	 */
	final private int _id;

	/**
	 * User defined name of the PowerPoint background
	 */
	final private String _name;
	
	/**
	 * Six character color for the font used with the background. (i.e. FFDDEE)
	 */
	final private String _fontColor;
	
	/**
	 * The original filename of the uploaded PowerPoint background 
	 */
	final private String _originalFilename;
	
	/**
	 * Filename of the PowerPoint background as it appears in storage
	 */
	final private String _filename;
	
	/**
	 * Filename of the background image thumbnail
	 */
	final private String _thumbnailFilename;
	
	/**
	 * UNIX timestamp of when the background was uploaded
	 */
	final private Timestamp _timestamp;
	
	/**
	 * Flags whether or not PowerPoint background has been archived. TRUE if PowerPoint background has been archived.
	 */
	final private boolean _archived;

	//---------------------------------------------------------------------------------------
	//------------------------ Constructors/Copiers/Destructors -----------------------------
	//---------------------------------------------------------------------------------------
	private PowerPointBackground(int backgroundID) throws SQLException {
		
		//Get the PowerPointBackground information from the database
		ResultSet backgroundResult = ASRDatabase.executeQuery("SELECT * " +
															  "FROM powerPointBackgrounds " + 
															  "WHERE PPB_id = " + backgroundID);		
		
		//If a PowerPointBackground was found with the passed ID, set the instance's members to its details
		if (backgroundResult.next()) {
			
			_id = backgroundResult.getInt("PPB_id");
			_name = backgroundResult.getString("PPB_name");
			_fontColor = backgroundResult.getString("PPB_fontColor");
			_originalFilename = backgroundResult.getString("PPB_originalFilename");
			_filename = backgroundResult.getString("PPB_filename");
			_thumbnailFilename = backgroundResult.getString("PPB_thumbnailFilename");
			_timestamp = backgroundResult.getTimestamp("PPB_timestamp");
			_archived = backgroundResult.getBoolean("PPB_archived");
		}	
		
		//Otherwise throw an error
		else {
			throw new AdShotRunnerException("Could not find PowerPointBackground with ID: " + backgroundID);
		}
	}
	
	//---------------------------------------------------------------------------------------
	//------------------------------- Modification Methods ----------------------------------
	//---------------------------------------------------------------------------------------
	//********************************* Public Methods **************************************		
	/**
	 * @return	Downloads the background image file and returns it (Should be deleted after use)
	 */
	public File file() {
		
		//If the background filename is empty, simply return null
		if (_filename.isEmpty()) {return null;}
		
		//Download the file
		try {
			URL backgroundURL = new URL(url());		
			String backgroundFilename = ASRProperties.pathForTemporaryFiles() + _filename;
			File backgroundFile = new File(backgroundFilename);		
			FileUtils.copyURLToFile(backgroundURL, backgroundFile);
	
			//Return the downloaded background file
			return backgroundFile;
		}
		catch (Exception e) {return null;}
	}
	

	//---------------------------------------------------------------------------------------
	//----------------------------------- Accessors -----------------------------------------
	//---------------------------------------------------------------------------------------
	//********************************* Public Accessors ************************************
	/**
	 * @return	ID of the PowerPoint background
	 */
	public int id() {return _id;}
	
	/**
	 * @return	User defined name of the PowerPoint background
	 */
	public String name() {return _name;}
	
	/**
	 * @return	Six character color for the font used with the background. (i.e. FFDDEE)
	 */
	public String fontColor() {return _fontColor;}
	
	/**
	 * @return	The original filename of the uploaded PowerPoint background
	 */
	public String originalFilename() {return _originalFilename;}
	
	/**
	 * @return	Filename of the PowerPoint background as it appears in storage
	 */
	public String filename() {return _filename;}
	
	/**
	 * @return	Full URL including protocol to the background image
	 */
	public String url() {
		if (_filename.isEmpty()) {return "";}
		return BACKGROUNDURLPATH + _filename;
	}
	
	/**
	 * @return	Filename of the background image thumbnail
	 */
	public String thumbnailFilename() {return _thumbnailFilename;}
	
	/**
	 * @return	Full URL including protocol to the background thumbnail image
	 */
	public String thumbnailURL() {
		if (_thumbnailFilename.isEmpty()) {return "";}
		return THUMBNAILURLPATH + _thumbnailFilename;
	}
	
	/**
	 * @return	UNIX timestamp of when the background was uploaded
	 */
	final public Timestamp timestamp() {return _timestamp;}
	
	/**
	 * @return	Flags whether or not PowerPoint background has been archived. TRUE if PowerPoint background has been archived.
	 */
	public boolean archived() {return _archived;}
	
}

File: frontend/classes/AdShotRunner/PowerPoint/PowerPointBackground.php
<?PHP
/**
* Contains the class for creating and retrieving PowerPoint backgrounds
*
* @package Adshotrunner
* @subpackage Classes
*/

namespace AdShotRunner\PowerPoint;

use AdShotRunner\System\ASRProperties;
use AdShotRunner\Database\ASRDatabase;
use AdShotRunner\Utilities\FileStorageClient;

/**
* The PowerPointBackground controls the flow of information between the system and the database 
* concerning PowerPoint background information and uploads the images to storage. 
*/
class PowerPointBackground {

	//---------------------------------------------------------------------------------------
	//---------------------------------- Constants ------------------------------------------
	//---------------------------------------------------------------------------------------
	//Dimensions of PowerPoint background thumbnail	
	const THUMBNAILMAXWIDTH = 200;
	const THUMBNAILMAXHEIGHT = 80;

	//URL path for images
	const IMAGEURLPATH = "https://s3.amazonaws.com/";

	//---------------------------------------------------------------------------------------
	//-------------------------------- Static Methods ---------------------------------------
	//---------------------------------------------------------------------------------------	
	//***************************** Public Static Methods ***********************************
	/**
	* Returns an instance of the PowerPointBackground with the passed ID.
	*
	* @param 	int 	$powerPointBackgroundID  	ID of the PowerPoint background to retrieve
	* @return 	PowerPointBackground				Instance of the PowerPointBackground to be retrieved. NULL on failure.
	*/
	static public function getPowerPointBackground($powerPointBackgroundID) {
		
		//Verify the ID and that it's not less than 1. If not, return NULL.
		if ((!$powerPointBackgroundID) || (!is_numeric($powerPointBackgroundID)) || ($powerPointBackgroundID < 1)) {return NULL;}
		
		//Get the info from the powerPointBackgrounds table
		$getBackgroundQuery = "SELECT * FROM powerPointBackgrounds WHERE PPB_id = $powerPointBackgroundID";
		$getBackgroundResult = ASRDatabase::executeQuery($getBackgroundQuery);
		$backgroundInfo = $getBackgroundResult->fetch_assoc();;

		//If no data was returned, return NULL.
		if (!$backgroundInfo) {return NULL;}
		
		//Create an instance of the class to return with the retrieved information
		$background = new PowerPointBackground();
		$background->_id = $powerPointBackgroundID;
		$background->_name = $backgroundInfo['PPB_name'];
		$background->_fontColor = $backgroundInfo['PPB_fontColor'];
		$background->_originalFilename = $backgroundInfo['PPB_originalFilename'];
		$background->_filename = $backgroundInfo['PPB_filename'];
		$background->_thumbnailFilename = $backgroundInfo['PPB_thumbnailFilename'];
		$background->_timestamp = strtotime($backgroundInfo['PPB_timestamp']);
		$background->_archived = $backgroundInfo['PPB_archived'];
		
		//Return the background instance
		return $background;
	}

	/**
	* Creates a PowerPointBackground with the passed background information and stores the image
	*
	* On success, the a PowerPointBackground is returned. On failure, NULL is returned.
	*
	* @param string 	$name   				Name of the PowerPoint background
	* @param string 	$fontColor 				Font color to be used witht the background
	* @param string 	$originalImageFilename  The original background image filename
	* @param string 	$imageFilename			Filename of the background image with associated path
	* @param int 	 	$userID 				ID of the user associated with the background
	* @return PowerPointBackground  			PowerPointBackground with the new information. NULL on failure.
	*/
	static public function create($name, $fontColor, $originalImageFilename, $imageFilename, $userID) {
		
		//Verify all parameters were passed.
		if ((!$name) || (!$fontColor) || (!$originalImageFilename) || 
			(!$imageFilename) || (!$userID)) {return NULL;}

		//Get the filename extension and return NULL if unavailable
		$imageType = self::getFilenameExtension($originalImageFilename);
		if (!$imageType) {return NULL;}

		//If the image extension type is not jpg, png, or bmp, return NULL
		if (($imageType != "jpg") && ($imageType != "jpeg") && ($imageType != "png") && ($imageType != "bmp")) {
			return NULL;
		}

		//Create the new filename: filename + user ID + timestamp + extension
		$namePart = explode(".", $originalImageFilename);
		$newFilename = $namePart[0] . "-" . $userID . "-" . time() . "." . $imageType;

		//Save the background image
		FileStorageClient::saveFile(ASRProperties::containerForPowerPointBackgrounds(), $imageFilename, $newFilename);

		//Create the thumbnail
		$thumbnailFilename = $namePart[0] . "-" . $userID . "-" . time() . "_thumbnail.png";
		self::createPNGThumbnail($imageFilename, RESTRICTEDPATH . 'temporaryFiles/' . $thumbnailFilename);

		//Save the thumbnail and delete the local file
		FileStorageClient::saveFile(ASRProperties::containerForPowerPointBackgrounds(), 
									RESTRICTEDPATH . 'temporaryFiles/' . $thumbnailFilename, 
									"thumbnails/" . $thumbnailFilename);
		unlink(RESTRICTEDPATH . 'temporaryFiles/' . $thumbnailFilename);
		
		//Add the info to the table
		$insertBackgroundQuery = "	INSERT INTO powerPointBackgrounds (	PPB_name,
																		PPB_fontColor,
																		PPB_originalFilename,
																		PPB_filename,
																		PPB_thumbnailFilename,
																		PPB_USR_id)
								 	VALUES ('" . ASRDatabase::escape($name) . "',
											'" . ASRDatabase::escape($fontColor) . "',
											'" . ASRDatabase::escape($originalImageFilename) . "',
											'" . ASRDatabase::escape($newFilename) . "',
											'" . ASRDatabase::escape($thumbnailFilename) . "',
											'" . ASRDatabase::escape($userID) . "')";
		ASRDatabase::executeQuery($insertBackgroundQuery);

		//Return the newly created PowerPointBackground
		return self::getPowerPointBackground(ASRDatabase::lastInsertID());
	}

	/**
	* Archives the PowerPoint background in the system with the passed ID.
	*
	* Archives the PowerPoint background that corresponds to the passed ID. On success, 
	* the PowerPointBackground object is returned. On failure, NULL is returned. 
	*
	* @param 	int 	$powerPointBackgroundID  	ID of the PowerPoint background to archive
	* @return 	PowerPointBackground  				Instance of the PowerPointBackground archived
	*/
	static public function archive($powerPointBackgroundID) {
		
		//Verify and ID was passed and it's not less than 1. If not, return NULL.
		if ((!$powerPointBackgroundID) || (!is_numeric($powerPointBackgroundID)) || ($powerPointBackgroundID < 1)) {return NULL;}
		
		//Verify the PowerPoint background in the database is not archived.
		$currentBackground = PowerPointBackground::getPowerPointBackground($powerPointBackgroundID);
		if ($currentBackground == NULL) {return NULL;}
		if ($currentBackground->isArchived()) {return $currentBackground;}
				
		//Archive the background
		$archiveBackgroundQuery = "	UPDATE powerPointBackgrounds 
							 	  	SET PPB_archived = 1
							 		WHERE PPB_id = " . $powerPointBackgroundID;
		ASRDatabase::executeQuery($archiveBackgroundQuery);
		
		//Send the archived background.
		return self::getPowerPointBackground($powerPointBackgroundID);
	}

	//***************************** Private Static Methods **********************************
	/**
	* Returns the extension of the filename.
	*
	* The extension is simply the text after the last '.' . This function returns it as lowercase.
	*
	* @param string 	$filename  		Filename
	* @return string 			 		Lowercase extension of the filename.
	*/	
	static private function getFilenameExtension($filename) {return strtolower(substr(strrchr($filename, "."), 1));}

	/**
	* Creates a PNG thumbnail of the passed image and saves it to the specified file.
	*
	* The thumbnail size is defined by the class constants THUMBNAILMAXWIDTH and THUMBNAILMAXHEIGHT
	*
	* @param string 	$imageFilename  		Image file to create thumbnail for
	* @param string 	$thumbnailFilename  	File to save thumbnail to
	* @return string 			 				Lowercase extension of the filename.
	*/	
	static private function createPNGThumbnail($imageFilename, $thumbnailFilename) {
	    $imagick = new \Imagick(realpath($imageFilename));
	    $imagick->thumbnailImage(self::THUMBNAILMAXWIDTH, self::THUMBNAILMAXHEIGHT, true);
	    $imagick->setImageFormat('png');
	    $imagick->writeImage($thumbnailFilename);
	}

	//---------------------------------------------------------------------------------------
	//------------------------------------ Variables ----------------------------------------
	//---------------------------------------------------------------------------------------	
	//******************************** Private Variables ************************************
	/**
	* @var int  Unique ID of the PowerPoint background
	*/
	private $_id;

	/**
	* @var string  User defined name of the PowerPoint background
	*/
	private $_name;	

	/**
	* @var string  Six character color for the font used with the background. (i.e. FFDDEE)
	*/
	private $_fontColor;

	/**
	* @var string  The original filename of the uploaded PowerPoint background
	*/
	private $_originalFilename;

	/**
	* @var string  Filename of the PowerPoint background as it appears in storage
	*/
	private $_filename;

	/**
	* @var string  Filename of the background image thumbnail
	*/
	private $_thumbnailFilename;

	/**
	* @var int	UNIX timestamp of when the background was uploaded
	*/
	private $_timestamp;

	/**
	* @var boolean	Flags whether or not PowerPoint background has been archived. TRUE if PowerPoint background has been archived.
	*/
	private $_archived;

	//---------------------------------------------------------------------------------------
	//------------------------ Constructors/Copiers/Destructors -----------------------------
	//---------------------------------------------------------------------------------------
	/**
	* Constructor set to private so only static factories can create new instances
	*/
	private function __construct() {}

	//---------------------------------------------------------------------------------------
	//----------------------------------- Accessors -----------------------------------------
	//---------------------------------------------------------------------------------------
	//********************************* Public Accessors ************************************
	/**
	* @return int  ID of the PowerPoint background
	*/
	public function id() {
		return $this->_id;
	}

	/**
	* @return string	User defined name of the PowerPoint background
	*/
	public function name() {
		return $this->_name;
	}

	/**
	* @return string	Font color to be used with the background
	*/
	public function fontColor() {
		return $this->_fontColor;
	}

	/**
	* @return string	Original filename associated with the background image
	*/
	public function originalFilename() {
		return $this->_originalFilename;
	}

	/**
	* @return string	Filename for the background image as it is in storage
	*/
	public function filename() {
		return $this->_filename;
	}

	/**
	* @return string	Full URL including protocol to the background image
	*/
	public function url() {
		if ($this->_filename == "") {return "";}
		return $this::IMAGEURLPATH . ASRProperties::containerForPowerPointBackgrounds() . "/" . $this->_filename;
	}

	/**
	* @return string	Filename of the background image's thumbnail
	*/
	public function thumbnailFilename() {
		return $this->_thumbnailFilename;
	}

	/**
	* @return string	Full URL including protocol to the background thumbnail image
	*/
	public function thumbnailURL() {
		if ($this->_thumbnailFilename == "") {return "";}
		return $this::IMAGEURLPATH . ASRProperties::containerForPowerPointBackgrounds() . "/thumbnails/" . $this->_thumbnailFilename;
	}

	/**
	* @return integer  Unix timestamp of when the background was uploaded
	*/
	public function timestamp() {
		return $this->_timestamp;
	}

	/**
	* @return boolean  Archive status of the PowerPoint background
	*/
	public function archived() {
		return $this->_archived;
	}


	//********************************* Private Accessors ***********************************




}